### PR TITLE
Tutorial for multi-label classification

### PR DESCRIPTION
**Pull request by aditya1503**
Tutorial for Multi-label classification

### CODE REVIEW

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    "plt.savefig(\"errors.pdf\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d65af827-aeda-4b6b-9ae7-b1f0b84700d5",
+   "metadata": {},
+   "source": [
+    "## Converting multilabels to onehot encoding?\n",
+    "In the multilabel setting, cleanlab expects labels to be a list of lists (length N)\n",
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:263]:
```suggestion
    "## Have labels formatted as a one-hot (multi-hot) binary matrix?\n",
    "For multi-label classification, cleanlab expects labels to be formatted as a list of lists, where each entry is an integer corresponding to a particular class. Here are some functions you can use to easily convert labels between these two formats. \n",
```

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    "```python\n",
+    "from cleanlab.internal.multilabel_utils import int2onehot\n",
+    "N = num_classes\n",
+    "y_one = int2onehot(labels, K=num_classes)\n",
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:267]:
```suggestion
    "\n",
    "y_binary_matrix = int2onehot(labels, K=num_classes)\n",
```

Comment (jwmueller) [reply docs/source/tutorials/multilabel_classification.ipynb:267]: not clear why you had `N = num_classes` here?

Comment (jwmueller) [reply docs/source/tutorials/multilabel_classification.ipynb:267]: You should also demonstrate converting label formats in the reverse direction, as many people originally start with their labels in the binary format.

```
labels_list_format = onehot2int(y_binary_matrix)
```

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 3]
```diff
@@ -0,0 +1,293 @@
+{
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:1]: Can we replace: "multilabel" with "multi-label" in all the text?

Comment (jwmueller) [reply docs/source/tutorials/multilabel_classification.ipynb:1]: and should generally say: "multi-label classification datasets", not just "multi-label datasets", except for where the context is clear.

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 2 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:278]: Curtis changed int2onthot to always include number of classes, hence I've added here

Comment (jwmueller) [reply docs/source/tutorials/multilabel_classification.ipynb]: sorry to clarify:  I wasn't asking about that, I was asking why there was `N = num_classes` line. I assume that was accidentally included, as I don't see `N` being used in the later code. If I am misunderstanding, would be helpful if you can explain what the role of `N` is (should generally not use this variable name for number of classes btw, always use `K` instead)

Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb]: ok got it, thanks for clarifying

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d65af827-aeda-4b6b-9ae7-b1f0b84700d5",
+   "metadata": {},
+   "source": [
+    "## Have labels formatted as a one-hot (multi-hot) binary matrix?\n",
+    "For multi-label classification, cleanlab expects labels to be formatted as a list of lists, where each entry is an integer corresponding to a particular class. Here are some functions you can use to easily convert labels between these two formats. \n",
+    "```python\n",
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:274]: Let's make this new code cell that is actually runnable

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "74ca37ea-cc7c-42f2-9330-dac806b015c3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "from cleanlab.filter import find_label_issues\n",
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:104]: let's import cleanlab functions after other imports in all tutorials

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    "import numpy as np\n",
+    "import os\n",
+    "import torchvision\n",
+    "\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "np.set_printoptions(suppress=True)"
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:111]:
```suggestion
    "\n",
    "np.set_printoptions(suppress=True)"
```

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    "<div  class=markdown markdown=\"1\" style=\"background:white;margin:16px\">  \n",
+    "    \n",
+    "```ipython3 \n",
+    "from cleanlab.filter import find_label_issues\n",
+    "\n",
+    "noise_idx = find_label_issues(\n",
+    "        labels=dataset[\"labels\"],\n",
+    "        pred_probs=dataset[\"pred_probs\"],\n",
+    "        multi_label=True,\n",
+    "        return_indices_ranked_by='normalized_margin',\n",
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:45]: How did you decide to use this option vs other options?

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    "    img = torchvision.io.read_image(path).swapaxes(0,2).swapaxes(0,1)\n",
+    "    fig.add_subplot(rows, columns, i)\n",
+    "    plt.axis('off')\n",
+    "    plt.imshow(img)\n",
+    "    plt.text(200, 50, data2[pos].split('/')[-1], bbox=dict(fill=False, edgecolor='black', linewidth=1))\n",
+    "    plt.text(200, 100, label_error, bbox=dict(fill=False, edgecolor='red', linewidth=1))\n",
+    "    plt.text(200, 150, pred_label_names[pos], bbox=dict(fill=False, edgecolor='green', linewidth=1))\n",
+    "    ct+=1\n",
+    "plt.savefig(\"errors.pdf\")"
+   ]
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:266]: Running this cell gives me error.

Screen Shot 2022-11-01 at 10 36 56 PM

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fffa88f6-84d7-45fe-8214-0e22079a06d1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "w = 10\n",
+    "h = 10\n",
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:247]: [nit] try to use more informative variable names

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.9.12"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 5
+}
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:304]: Don't worry about polishing the text too much, I'll do a rewrite of the text for you to make the tutorial as clear as possible

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 8 above)
Comment (jwmueller) [reply docs/source/tutorials/multilabel_classification.ipynb:266]: probably your wget call did not download data to the right location

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    "```ipython3\n",
+    "!pip install cleanlab\n",
+    "!pip install torchvision\n",
+    "\n",
+    "\n",
+    "# Make sure to install the version corresponding to this tutorial\n",
+    "# E.g. if viewing master branch documentation:\n",
+    "#     !pip install git+https://github.com/cleanlab/cleanlab.git\n",
+    "\n",
+    "```\n",
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:70]: Your notebook is missing a critical cell needed to make this work on Colab:

See any of our other tutorial notebooks for an example, eg:
https://github.com/cleanlab/cleanlab/blob/master/docs/source/tutorials/outliers.ipynb

The missing cell is the one that starts with:
`# Package installation (hidden on docs website).`

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+    "## Have labels formatted as a one-hot (multi-hot) binary matrix?\n",
+    "For multi-label classification, cleanlab expects labels to be formatted as a list of lists, where each entry is an integer corresponding to a particular class. Here are some functions you can use to easily convert labels between these two formats. \n",
+    "```python\n",
+    "from cleanlab.internal.multilabel_utils import int2onehot\n",
+    "\n",
+    "y_binary_matrix = int2onehot(labels, K=num_classes)\n",
+    "```"
+   ]
+  }
+ ],
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:282]: Check out the final cell in this other tutorial:
https://github.com/cleanlab/cleanlab/blob/master/docs/source/tutorials/image.ipynb
(the one starting with `# Note: This cell is only for docs.cleanlab.ai, if running on local Jupyter or Colab, please ignore it.`)

Can we add a cell like that here? Note these are hidden cells which do not appear in the version of the notebook rendered on docs.cleanlab.ai, because their metadata contains a hidden tag (look at the raw .ipynb file in text editor to see this).  The idea is such a cell can serve as a regression test for your code, by ensuring `find_label_issues` returns some indices we expect to see in the top-ranked issues.

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+   "execution_count": null,
+   "id": "29ab0d31-5002-4935-8f96-3847edbc172c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "multi_labels = convert_multilabel_y_to_list(labels)\n",
+    "\n",
+    "pred_labels = convert_multilabel_y_to_list(np.array((pred_probs>0.5),np.int32))"
+   ]
+  },
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:190]: show examples of ~3 well-labeled images from CelebA here (say first 3 in the dataset)

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 7 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:45]: I tried other return_indices_ranked_by methods, this seemed to have more errors within top 10 issues (indices)

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 2 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 4 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 5 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0790e633-dd4b-4217-b32f-3e33ed56e739",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "labels=(df[class_names].to_numpy())"
```
[the remark this answers is not in the document]
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 8 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:361]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 11 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:70]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 12 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:384]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 13 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:272]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 9 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb (hunk 3 above)
Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:1]: Fixed: https://github.com/cleanlab/cleanlab/pull/517/commits/7aa53405e8abc5f3f0aa2f2a3a96f90f21516db1

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 15]
```diff
@@ -0,0 +1,406 @@
...
+   "id": "7383d024-8273-4039-bccd-aab3020d331f",
+   "metadata": {
+    "nbsphinx": "hidden"
+   },
+   "outputs": [],
+   "source": [
+    "# Package installation (hidden on docs website).\n",
+    "# Package versions we used: matplotlib==3.5.1, torch==1.11.0, torchvision==0.12.0, timm==0.5.4\n",
+    "\n",
+    "dependencies = [ \"torchvision\", \"scikit-learn\"]\n",
```
Comment (jwmueller) [inline docs/source/tutorials/multilabel_classification.ipynb:86]: matplotlib, timm, torch should be in this list?

Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:86]: I thought requirements.txt would cover, will add

Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb:86]: Just Matplotlib, this is the tutorial notebook, we don't need other libraries

Comment (aditya1503) [reply docs/source/tutorials/multilabel_classification.ipynb]: Added: https://github.com/cleanlab/cleanlab/pull/517/commits/d90d29e130901bbe1920df31d58a0102ccd2caf2

**Review by jwmueller — APPROVED** [review]
Made another final pass of text/code edits, so please review these and give me thumbs up before I merge this.

**Comment by jwmueller** [comment]
TODOs after I merge this in follow-up PR:

- [ ] make this tutorial visible on docs website by adding the tutorial to the appropriate index.rst pages as described in DEVELOPMENT.md

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 16]
```diff
@@ -0,0 +1,576 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "144ad4c2-49bb-4147-a743-a83ed1656a11",
+   "metadata": {},
+   "source": [
+    "## 2. Format data, labels, and model predictions\n",
+    "\n",
+    "In multi-label classification, each example in the dataset is labeled as belonging to one **or more** of *K* possible classes. To find label issues, cleanlab requires predicted class probabilities from a trained classifier. \n",
```
Comment (elisno) [inline docs/source/tutorials/multilabel_classification.ipynb:400]: In principle, this should be "zero or more".

Comment (jwmueller) [reply docs/source/tutorials/multilabel_classification.ipynb:400]: good catch! clarified

**Review by elisno — APPROVED** [review]
LGTM

**Comment by aditya1503** [comment]
> Made another final pass of text/code edits, so please review these and give me thumbs up before I merge this.

LGTM.
